skbeam/core/correlation.py

```diff
@@ -405,3 +409,356 @@ def auto_corr_scat_factor(lags, beta, relaxation_rate, baseline=1):
+                                                       labels, num_frames)
+    # create a shorthand reference to the results and state named tuple
+    s = two_time_internal_state
+
+    for img in images:
+        s.cur[0] = (1 + s.cur[0]) % num_bufs  # increment buffer
+
+        s.count_level[0] = 1 + s.count_level[0]
+
+        # get the current image time
+        s = s._replace(current_img_time=(s.current_img_time + 1))
+
+        # Put the image into the ring buffer.
+        s.buf[0, s.cur[0] - 1] = (np.ravel(img))[s.pixel_list]
+
+        # Compute the two time correlations between the first level
+        # (undownsampled) frames. two_time and img_per_level in place!
+        _two_time_process(s.buf, s.two_time, s.label_array, num_bufs,
+                          s.num_pixels, s.img_per_level, s.lag_steps,
+                          s.current_img_time, level=0, buf_no=s.cur[0] - 1)
+
+        # time frame for each level
+        s.time_ind[0].append(s.current_img_time)
+
+        # check whether the number of levels is one, otherwise
+        # continue processing the next level
+        processing = num_levels > 1
+
+        # Compute the correlations for all higher levels.
+        level = 1
+        while processing:
+            if not s.track_level[level]:
+                s.track_level[level] = 1
+                processing = False
+            else:
+                prev = 1 + (s.cur[level - 1] - 2) % num_bufs
+                s.cur[level] = 1 + s.cur[level] % num_bufs
+                s.count_level[level] = 1 + s.count_level[level]
+
+                s.buf[level, s.cur[level] - 1] = (s.buf[level - 1, prev - 1] +
+                                                  s.buf[level - 1,
+                                                  s.cur[level - 1] - 1])/2
+
+                t1_idx = (s.count_level[level] - 1) * 2
+
+                current_img_time = ((s.time_ind[level - 1])[t1_idx]
+                                    + (s.time_ind[level - 1])[t1_idx + 1])/2.
+
+                # time frame for each level
+                s.time_ind[level].append(current_img_time)
+
+                # make the track_level zero once that level is processed
+                s.track_level[level] = 0
+
+                # call the _two_time_process function for each multi-tau level
+                # for multi-tau levels greater than one
+                # Again, this is modifying things in place. See comment
+                # on previous call above.
+                _two_time_process(s.buf, s.two_time, s.label_array, num_bufs,
+                                  s.num_pixels, s.img_per_level, s.lag_steps,
+                                  current_img_time, level=level,
+                                  buf_no=s.cur[level]-1)
+                level += 1
+
+                # Checking whether there is next level for processing
+                processing = level < num_levels
+
+    for q in range(np.max(s.label_array)):
```

Comment:
This generator implementation is different from `lazy_one_time`.  You should `yield s` after **every** image and then provide a helper function to convert `s` into `g2 and `lag_steps`.  Lines 557-568 should be converted into something like this:

```python
                level += 1

                # Checking whether there is next level for processing
                processing = level < num_levels
        yield s

def two_time_state_to_results(state):
    """Convert the internal state of the two time generator into usable results

    Parameters
    ----------
    state : namedtuple
        The internal state that is yielded from `lazy_two_time`

    Returns
    -------
    results : namedtuple
        A results object that contains the two time correlation results and the
        lag steps
    """
    for q in range(np.max(state.label_array)):
        x0 = (state.two_time)[:, :, q]
        (state.two_time)[:, :, q] = (np.tril(x0) + np.tril(x0).T -
                                     np.diag(np.diag(x0)))
    return results(state.two_time, state.lag_steps, state)
```

You will likely need to update the tests accordingly.  

It would also make sense to remove `state` from the `results` namedtuple, but that would then require making similar changes to the `lazy_one_time` generator and all code paths, so you are welcome to punt on doing that in this PR and I can do it later.